### PR TITLE
parser: Fix handling of multiple left angles in null denotation

### LINKED ISSUE #1880: Invalid parsing of complex qualified path in expression
I tried this code:

```rust
fn main() {
    let _ = <<F as Foo>::A as Bar>::f();
}
```

I expected to see this happen: no parsing error. `rustc` with `-Z parse-only` returns 0

Instead, this happened: 

```rust
test.rs:2:13: error: found unexpected token ‘<<’ in null denotation
    2 |     let _ = <<F as Foo>::A as Bar>::f();
      |             ^~
test.rs:2:15: error: failed to parse expression in let statement
    2 |     let _ = <<F as Foo>::A as Bar>::f();
      |               ^
test.rs:2:5: error: failed to parse statement or expression without block in block expression
    2 |     let _ = <<F as Foo>::A as Bar>::f();
      |     ^~~
test.rs:3:1: error: unrecognised token ‘}’ for start of item
    3 | }
      | ^
test.rs:3:1: error: failed to parse item in crate
```

or

```rust
test.rs:2:35: error: found null denotation scope resolution operator, and have not written handling for it
    2 |     let _ = <<F as Foo>::A as Bar>::f();
      |                                   ^~
test.rs:2:37: error: failed to parse expression in let statement
    2 |     let _ = <<F as Foo>::A as Bar>::f();
      |                                     ^
test.rs:2:5: error: failed to parse statement or expression without block in block expression
    2 |     let _ = <<F as Foo>::A as Bar>::f();
      |     ^~~
test.rs:3:1: error: unrecognised token ‘}’ for start of item
    3 | }
      | ^
test.rs:3:1: error: failed to parse item in crate
```

with basic handling of `<<` in `null_denotation`

### Meta

- What version of Rust GCC were you using: 1bd1e466453

### PR DESCRIPTION
**Pull request by goar5670**
Fixes #1880

### CODE REVIEW

File: gcc/testsuite/rust/compile/complex_qualified_path_in_expr.rs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// { dg-additional-options "-fsyntax-only" }
```
Comment (philberty) [inline gcc/testsuite/rust/compile/complex_qualified_path_in_expr.rs:1]: have you tried this without -fsyntax-only? I merged in https://github.com/Rust-GCC/gccrs/pull/1954 which fixes issues there

**Review by philberty — CHANGES_REQUESTED** [review]
I would like to see the test-case updated to not use -fsyntax-only or if there is an issue for an issue to be raised

File: gcc/testsuite/rust/compile/complex_qualified_path_in_expr.rs (hunk 1 above)
Comment (goar5670) [reply gcc/testsuite/rust/compile/complex_qualified_path_in_expr.rs:1]: Good point. I tried it as per your suggestion and it exposed a couple of other issues.
1. `let _ = <<S as T1>::A as T2>::f()` failing: This was caused by `rust-hir-path.h`'s `gcc_unreachable ()` at line 948 in combination with the missing initialization for `associated_segment` in `HIR::QualifiedPathInType`'s constructor (`rust-hir-path.h` line 941) I'm guessing it wasn't intentional, right?
2. After fixing 1, `let _ = <<<S as T1>::A as T2>::B as T3>::f();` failed: This I couldn't solve completely but I'll share what I understood. `ResolveRelativeQualTypePath::resolve_qual_seg` expects a resolved node id to be returned which is not applicable if the the `Type` (as in reference to < Type as TypePath >) in the `QualifiedPathType` is a `QualifiedPathInType`. I'm not sure what the correct way of handling this is but would like to hear your input.

**Review by philberty — APPROVED** [review]
LGTM

**Pull request merged**